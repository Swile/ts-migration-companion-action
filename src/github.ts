import {context, getOctokit} from '@actions/github'

export const isGithubBotComment = (comment: {
  user?: {login: string} | null
}): boolean => comment.user?.login === 'github-actions[bot]'

export const createOrReplaceGithubBotComment = async (
  githubClient: ReturnType<typeof getOctokit>,
  message: string
): Promise<void> => {
  const contextArgs = {
    owner: context.repo.owner,
    repo: context.repo.repo,
    issue_number: context.issue.number
  }

  const response = await githubClient.rest.issues.listComments(contextArgs)
  const comments = response.status === 200 ? response.data : []

  await Promise.all(
    comments.map(async comment => {
      if (isGithubBotComment(comment)) {
        return githubClient.rest.issues.deleteComment({
          ...contextArgs,
          comment_id: comment.id
        })
      }
      return Promise.resolve(null)
    })
  )

  await githubClient.rest.issues.createComment({
    ...contextArgs,
    body: message
  })
}
