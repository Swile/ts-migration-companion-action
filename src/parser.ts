import path from 'node:path'

export function normalizeFilepath(filepath: string): {
  basepath: string
  extension: string
} {
  const extension = path.extname(filepath)
  const basepath = path.join(
    path.dirname(filepath),
    path.basename(filepath, extension)
  )
  return {
    basepath,
    extension
  }
}

export interface ParsedTsError {
  basepath: string
  extension: string
  position: string
  error: string
  code: string
  message: string
}

export function parseTscOutput(
  output: string,
  include: string[]
): ParsedTsError[] {
  const ast = output
    .split('\n')
    .filter(line => include.some(prefix => line.startsWith(prefix)))
    .map(parseErrorLine)

  return ast
}

export function parseErrorLine(line: string): ParsedTsError {
  const filepath = line.substring(0, line.indexOf('('))
  const position = line.substring(line.indexOf('(') + 1, line.indexOf(')'))
  const error = line.substring(line.indexOf(': error ') + ': error '.length)
  const code = error.substring(0, error.indexOf(': '))
  const message = error.substring(error.indexOf(': ') + 2)

  const {basepath, extension} = normalizeFilepath(filepath)

  return {basepath, extension, position, error, code, message}
}
