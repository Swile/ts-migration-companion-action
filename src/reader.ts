import fs from 'node:fs/promises'

export async function readTscOutputByFilepath(
  filepath: string
): Promise<string> {
  return await fs.readFile(filepath, {encoding: 'utf8'})
}
