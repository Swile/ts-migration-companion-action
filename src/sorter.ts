interface Countable {
  count: number
}

export function sortByCount(left: Countable, right: Countable): number {
  return right.count - left.count
}
