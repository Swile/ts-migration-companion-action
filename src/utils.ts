import {ReadStream} from 'fs'

export async function streamToString(stream: ReadStream): Promise<string> {
  const chunks: Buffer[] = []
  return new Promise((resolve, reject) => {
    stream.on('data', (chunk: Buffer) => chunks.push(chunk))
    stream.on('error', reject)
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')))
  })
}
