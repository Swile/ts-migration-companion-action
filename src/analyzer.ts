import {type ParsedTsError} from './parser'

export interface TsReport {
  [basepath: string]: string[]
}

export interface TsErrorByFile {
  basepath: string
  count: number
}

function groupByFileReport(
  report: TsReport,
  {basepath, code}: ParsedTsError
): TsReport {
  const byFileBasePath = [...(report[basepath] ?? []), code]
  report[basepath] = byFileBasePath
  return report
}

function combineReportByFilename(report: TsReport): TsErrorByFile[] {
  return Object.entries(report).map(([basepath, errors]) => ({
    basepath,
    count: errors.length
  }))
}

export function getErrorByFileReport(ast: ParsedTsError[]): TsErrorByFile[] {
  const report: TsReport = ast.reduce(groupByFileReport, {})
  return combineReportByFilename(report)
}
