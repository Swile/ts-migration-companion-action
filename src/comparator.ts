/* eslint-disable github/array-foreach */
import {getErrorByFileReport} from './analyzer'
import {getMarkdownReportComparator, type MergedTsErrorByFile} from './report'
import {ParsedTsError, parseTscOutput} from './parser'
import {readTscOutputByFilepath} from './reader'

type ComparatorArgs = {
  masterTscOutputPath: string
  prTscOutputPath: string
  tscRootDir: string
}

async function comparator({
  masterTscOutputPath,
  prTscOutputPath,
  tscRootDir
}: ComparatorArgs): Promise<string> {
  const masterTscOutput = await readTscOutputByFilepath(masterTscOutputPath)
  const prTscOutput = await readTscOutputByFilepath(prTscOutputPath)

  const masterAst = parseTscOutput(masterTscOutput, [tscRootDir])
  const prAst = parseTscOutput(prTscOutput, [tscRootDir])

  const masterReport = getErrorByFileReport(masterAst)
  const prReport = getErrorByFileReport(prAst)

  const mergedReport: MergedTsErrorByFile[] = []

  masterReport.forEach(masterError => {
    const prError = prReport.find(
      item => item.basepath === masterError.basepath
    )
    mergedReport.push({
      basepath: masterError.basepath,
      master: masterError.count,
      pr: prError?.count ?? 0
    })
  })

  prReport.forEach(prError => {
    const masterError = prReport.find(
      ({basepath}) => basepath === prError.basepath
    )
    const mergedError = mergedReport.find(
      ({basepath}) => basepath === prError.basepath
    )
    if (mergedError) {
      // Skip already merged errors
      return
    }

    mergedReport.push({
      basepath: prError.basepath,
      master: masterError?.count ?? 0,
      pr: prError.count
    })
  })

  // Files without error in master
  const openedList = mergedReport.filter(merged => {
    return merged.master === 0 && merged.pr > 0
  })

  // Files without error in pr
  const fixedList = mergedReport.filter(merged => {
    return merged.pr === 0 && merged.master > 0
  })

  // Files with error count higher in pr than master
  const increasedList = mergedReport.filter(merged => {
    return merged.pr > merged.master && merged.master > 0
  })

  // Detailed list of tsc errors
  const increasedDetails = increasedList.reduce((map, error) => {
    const details = prAst.filter(value => value.basepath === error.basepath)
    map.set(error.basepath, details)
    return map
  }, new Map<string, ParsedTsError[]>())

  // Files with error count lower in pr than master
  const decreasedList = mergedReport.filter(merged => {
    return merged.pr < merged.master && merged.pr > 0
  })

  return getMarkdownReportComparator({
    global: {
      master: masterAst.length,
      pr: prAst.length
    },
    openedList,
    fixedList,
    increasedList,
    increasedDetails,
    decreasedList
  })
}

export default comparator
