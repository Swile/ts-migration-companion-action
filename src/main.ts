import * as core from '@actions/core'
import {getOctokit} from '@actions/github'

import comparator from './comparator'
import {createOrReplaceGithubBotComment} from './github'

export const getActionInputs = (): {
  token: string
  masterTscOutputPath: string
  prTscOutputPath: string
  tscRootDir: string
} => ({
  token: core.getInput('github-token', {required: true}),
  masterTscOutputPath: core.getInput('master_tsc_output_path', {
    required: true,
    trimWhitespace: true
  }),
  prTscOutputPath: core.getInput('pr_tsc_output_path', {
    required: true,
    trimWhitespace: true
  }),
  tscRootDir: core.getInput('tsc_rootdir')
})

export async function run(): Promise<void> {
  try {
    const {token, ...comparatorOptions} = getActionInputs()

    const markdown = await comparator(comparatorOptions)

    const github = getOctokit(token)

    await createOrReplaceGithubBotComment(github, markdown)
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

if (require.main === module) {
  run()
}
