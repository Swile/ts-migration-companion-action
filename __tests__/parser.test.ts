import {normalizeFilepath, parseErrorLine, parseTscOutput} from '../src/parser'

const output = `
src/api/v0/serializers/utils/serialize-types.js(4,26): error TS7006: Parameter 'value' implicitly has an 'any' type.
src/api/v0/serializers/utils/serialize-types.js(17,33): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '<T>(obj: T) => Pick<T, Exclude<keyof T, Exclude<keyof T, string | number | symbol>>>' is not assignable to parameter of type '(x0: any, x1: any, x2: any) => readonly any[]'.
      Type 'Pick<any, string | number | symbol>' is missing the following properties from type 'readonly any[]': length, concat, join, slice, and 18 more.
src/api/v0/serializers/utils/serialize-types.js(17,46): error TS2554: Expected 0 arguments, but got 1.
src/api/v0/serializers/utils/serialize-types.js(20,24): error TS7006: Parameter 'date' implicitly has an 'any' type.
src/api/v0/serializers/utils/serialize-types.js(23,31): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '<T>(obj: T) => Pick<T, Exclude<keyof T, Exclude<keyof T, string | number | symbol>>>' is not assignable to parameter of type '(x0: any, x1: any, x2: any) => readonly any[]'.
src/api/v0/serializers/utils/serialize-types.js(23,44): error TS2554: Expected 0 arguments, but got 1.`

describe('Parser', () => {
  describe('#normalizeFilepath', () => {
    it('should return basepath and file extension from filepath', () => {
      expect(
        normalizeFilepath('src/api/v0/serializers/utils/serialize-types.js')
      ).toEqual({
        basepath: 'src/api/v0/serializers/utils/serialize-types',
        extension: '.js'
      })
    })
  })

  describe('#parseErrorLine', () => {
    it('should extract informations from error line', () => {
      expect(parseErrorLine(output.split('\n')[1])).toEqual({
        basepath: 'src/api/v0/serializers/utils/serialize-types',
        code: 'TS7006',
        error: "TS7006: Parameter 'value' implicitly has an 'any' type.",
        extension: '.js',
        message: "Parameter 'value' implicitly has an 'any' type.",
        position: '4,26'
      })
    })
  })

  describe('#parseTscOutput', () => {
    it('correctly parse tsc output', () => {
      const ast = parseTscOutput(output, ['src'])
      expect(ast).toMatchSnapshot()
    })
  })
})
