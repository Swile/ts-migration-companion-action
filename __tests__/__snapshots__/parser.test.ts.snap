// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser #parseTscOutput correctly parse tsc output 1`] = `
[
  {
    "basepath": "src/api/v0/serializers/utils/serialize-types",
    "code": "TS7006",
    "error": "TS7006: Parameter 'value' implicitly has an 'any' type.",
    "extension": ".js",
    "message": "Parameter 'value' implicitly has an 'any' type.",
    "position": "4,26",
  },
  {
    "basepath": "src/api/v0/serializers/utils/serialize-types",
    "code": "TS2769",
    "error": "TS2769: No overload matches this call.",
    "extension": ".js",
    "message": "No overload matches this call.",
    "position": "17,33",
  },
  {
    "basepath": "src/api/v0/serializers/utils/serialize-types",
    "code": "TS2554",
    "error": "TS2554: Expected 0 arguments, but got 1.",
    "extension": ".js",
    "message": "Expected 0 arguments, but got 1.",
    "position": "17,46",
  },
  {
    "basepath": "src/api/v0/serializers/utils/serialize-types",
    "code": "TS7006",
    "error": "TS7006: Parameter 'date' implicitly has an 'any' type.",
    "extension": ".js",
    "message": "Parameter 'date' implicitly has an 'any' type.",
    "position": "20,24",
  },
  {
    "basepath": "src/api/v0/serializers/utils/serialize-types",
    "code": "TS2769",
    "error": "TS2769: No overload matches this call.",
    "extension": ".js",
    "message": "No overload matches this call.",
    "position": "23,31",
  },
  {
    "basepath": "src/api/v0/serializers/utils/serialize-types",
    "code": "TS2554",
    "error": "TS2554: Expected 0 arguments, but got 1.",
    "extension": ".js",
    "message": "Expected 0 arguments, but got 1.",
    "position": "23,44",
  },
]
`;
